db/migrate/20150625220141_fix_serialized_reports_for_rails_four.rb

```diff
@@ -126,34 +121,74 @@ def parts
 
   def up
     say_with_time("Converting MiqReportResult#report to a serialized hash") do
-      MiqReportResult.where('report IS NOT NULL').find_each do |rr|
-        val = rr.serialize_report_to_hash(rr.read_attribute(:report), self)
-        rr.update_attribute(:report, val) if val
+      base_relation = MiqReportResult.where('report IS NOT NULL')
+
+      say_batch_started(base_relation.size)
+
+      base_relation.find_in_batches do |batch|
+        MiqReportResult.transaction do
+          batch.each do |rr|
+            val = Serializer.serialize_report_to_hash(rr.report, rr.class, rr.id, self)
+            rr.update_attribute(:report, val) if val
+          end
+        end
+        say_batch_processed(batch.size)
       end
     end
 
     say_with_time("Converting BinaryBlob report results to a serialized hash") do
-      BinaryBlob.includes(:resource).where(:resource_type => 'MiqReportResult').find_each do |bb|
-        if bb.resource
-          val = bb.serialize_report_to_hash(bb.binary, self)
-          bb.binary = val if val
+      # Pre-autoload classes that appear in reports
+      MiqExpression
+      Ruport
+
+      all_ids = BinaryBlob
+                  .where(:resource_type => 'MiqReportResult')
+                  .select(:id, :resource_id, :resource_type)
+                  .includes(:resource)
+                  .select(&:resource) # Ignore orphaned BinaryBlob records
+                  .map(&:id)
+
+      say_batch_started(all_ids.size)
+
+      all_ids.each_slice(1000) do |ids|
+        bbs = BinaryBlob.where(:id => ids)
+
+        # Extract the binaries from the AR objects so we don't have to pass
+        #   the AR objects through to Parallel
+        MiqPreloader.preload(bbs, :binary_blob_parts)
+        payload = bbs.map { |bb| [bb.binary, bb.class.name, bb.id] }
+
+        # Convert the binaries in parallel processes
+        converted = Parallel.map(payload) do |binary, klass, id|
```

Comment:
I'm not sure what you mean.  The purpose of my changes are to break the outer BinaryBlob loop into batches of 1000 (so we are only holding 1000 records worth of BinaryBlobs in memory to keep the memory levels reasonable), and then I use `parallel` to process that 1000 records in parallel (because YAML conversion is the slow part and is CPU bound, this allows us to use multiple cores to do the hard work).